### PR TITLE
Integrate UrbanSoundDataset for Audio Data Processing

### LINKED ISSUE #1156: [FEATURE]: Integrate UrbanSoundDataset for Audio Trainspace
### Feature Name

As part of our initiative to extend the Deep Learning Playground's capabilities to include audio data processing, we need to integrate the UrbanSoundDataset into our system.

### Your Name

Surya Subramanian

## Description

This task involves enhancing the `training/core/dataset.py` module to support audio data, specifically focusing on the UrbanSound8K dataset. The goal is to create a seamless pipeline for data ingestion, preprocessing, and loading, tailored for audio files.

### Objectives

Take inspiration from the other classes in `dataset.py` and develop `dataCreator`, `train_loader`, and `test_loader` methods to facilitate the loading of preprocessed audio data into the model for training and testing. Ensure that data loading is efficient and integrates well with PyTorch's `DataLoader` mechanism. Essentially, the code in this class will be used to create train and test datasets. 

### Implementation Details
Your best bet is to take inspiration from ImageDefaultDatasetCreator and "recreate" that but for the UrbanSoundDataset. It should be seamless for us to call the createTrainDataset() and createTestDataset() methods from other files (like how it is being used called in `training/training/routes/image/image.py`.  

So in summary,
- The `UrbanSoundDataset` class should be added to `training/core/dataset.py`, encapsulating all functionalities related to the UrbanSound8K dataset.
- Ensure that the dataset loader is compatible with our existing training pipeline, allowing for smooth integration with models and training routines.
- UPDATE: schemas.py will be handled here. Here's a high level overview of what we are thinking:

```
# no layer params

# hardcode non tunable params in

# class AudioParams(Schema):
#     name: str
#     problem_type: Literal["CLASSIFICATION"]
#     default: Urban Sound Dataset
#     criterion: negative log likelihood loss
#     optimizer_name: Adam 
#     shuffle: bool
#     epochs: int
#     test_size: float
#     batch_size: int
#     user_arch: arch we are using 
```

Feel free to play around with this!

### PR DESCRIPTION
**Pull request by codingwithsurya**
# adding urbansound-dataset and schemas.py

Github Issue Number Here: <ntegrate UrbanSoundDataset for Audio Trainspace #1156>
**What user problem are we solving?**
We are enhancing the Deep Learning Playground's capabilities to include audio data processing by integrating the UrbanSound8K dataset. This allows users to work with audio data seamlessly within the existing pipeline, expanding the versatility and application of the platform.

**What solution does this PR provide?**
This PR adds a new class, UrbanSoundDataset, to the training/core/dataset.py module. This class encapsulates functionalities for data ingestion, preprocessing, and loading specifically tailored for the UrbanSound8K dataset. It includes dataCreator, train_loader, and test_loader methods to facilitate efficient data loading into the model for training and testing. Additionally, it ensures compatibility with PyTorch's DataLoader mechanism and integrates smoothly with the existing training pipeline.

It also provides a schemas.py file that provides audio params. This file is still a WIP. 

**Testing Methodology**
> How did you test your changes and verify that existing functionality is not broken
manual testing

**Any other considerations**
Updated schemas.py to include AudioParams for defining non-tunable parameters specific to the UrbanSound8K dataset.

we also added 2 new dependencies to poetry -- torchaudio and soundata

### CODE REVIEW

File: training/training/core/dataset.py [hunk 1]
```diff
@@ -201,3 +203,80 @@ def createTestDataset(self) -> DataLoader:
...
+        # Downsample the audio to ~8kHz
+        tempData = torch.zeros([160000, 1])  # tempData accounts for audio clips that are too short
+        if sound.numel() < 160000:
+            tempData[:sound.numel()] = sound[:]
+        else:
+            tempData[:] = sound[:160000]
+        
+        soundData = tempData
+        soundFormatted = torch.zeros([32000, 1])
+        soundFormatted[:32000] = soundData[::5]  # Take every fifth sample of soundData
```
Comment (karkir0003) [inline training/training/core/dataset.py:249]: explain?

**Comment by codingwithsurya** [comment]
looks like one of the lints are failing. weird. @karkir0003 @DSGT-DLP/project-lead 

the error message is this:
Note: This error originates from the build backend, and is likely not a problem with poetry but with simpleaudio (1.0.4) not supporting PEP 517 builds. You can verify this by running 'pip wheel --no-cache-dir --use-pep517 "simpleaudio (==1.0.4)"'.
Error: Process completed with exit code 1.

We just added torchaudio and soundata libraries in this pr to poetry so bc of that.

File: training/training/core/dataset.py [hunk 2]
```diff
@@ -201,3 +203,80 @@ def createTestDataset(self) -> DataLoader:
...
+        
+        # Convert to mono if stereo. Otherwise, just take the first channel. 
+        # we need this bc the UrbanSound8k dataset is stereo, but torchaudio defaults to mono
+        if sound.size(0) > 1:
+            sound = torch.mean(sound, dim=0, keepdim=True)
+        
+        # Normalize the audio. Mean and std are calculated from the training set
+        sound = (sound - sound.mean()) / sound.std()
+
+        # Downsample the audio to ~8kHz
```
Comment (karkir0003) [inline training/training/core/dataset.py:240]: did the paper do these preprocessing steps?

File: training/training/core/dataset.py [hunk 3]
```diff
@@ -201,3 +203,80 @@ def createTestDataset(self) -> DataLoader:
...
+        return soundFormatted, self.labels[index]
+    
+    def __len__(self):
+        return len(self.file_names)
+
+class UrbanSoundDatasetCreator(TrainTestDatasetCreator):
+    def __init__(self, folderList_train, folderList_test, batch_size=128, shuffle=True):
+        # Initialize and download the dataset using soundata
+        self.dataset = soundata.initialize('urbansound8k')
+        self.dataset.download()  # Download the dataset
```
Comment (karkir0003) [inline training/training/core/dataset.py:260]: will this actually create a csv file or smthing in the directory?

**Review by karkir0003 — COMMENTED** [review]
good start for the most part.

please address my questions and add unit tests to ensure the data loading logic works as intended

**Comment by karkir0003** [comment]
> looks like one of the lints are failing. weird. @karkir0003 @DSGT-DLP/project-lead 
> 
> the error message is this:
> Note: This error originates from the build backend, and is likely not a problem with poetry but with simpleaudio (1.0.4) not supporting PEP 517 builds. You can verify this by running 'pip wheel --no-cache-dir --use-pep517 "simpleaudio (==1.0.4)"'.
> Error: Process completed with exit code 1.
> 
> We just added torchaudio and soundata libraries in this pr to poetry so bc of that. 

whats simpleaudio? is this a lib thats a dependency used by soundata?

**Comment by karkir0003** [comment]
did we install the latest stable version of torchaudio and soundata?

**Comment by karkir0003** [comment]
> looks like one of the lints are failing. weird. @karkir0003 @DSGT-DLP/project-lead
> 
> the error message is this: Note: This error originates from the build backend, and is likely not a problem with poetry but with simpleaudio (1.0.4) not supporting PEP 517 builds. You can verify this by running 'pip wheel --no-cache-dir --use-pep517 "simpleaudio (==1.0.4)"'. Error: Process completed with exit code 1.
> 
> We just added torchaudio and soundata libraries in this pr to poetry so bc of that.

I'd try starting the debugging with the following:
1. Try running the command that's shown in the log. See if any further logs come up
2. Maybe we might need to find a compatible version of `soundata` or potentially add `simpleaudio` as a dependency just like how you used DLP CLI to install `torchaudio`
3. If 1 and 2 don't work, try asking in the poetry github repo by filing a github issue. There's also a discord server for Poetry that can help clarify

File: training/training/core/dataset.py (hunk 1 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py:249]: the reason why we are taking every fifth sample from soundData and assigning it to the first 32000 elements of soundFormatted is to downsample the data if soundData is a high-frequency sound signal.

File: training/training/core/dataset.py (hunk 2 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py:240]: yes. this is just simple normalization.

File: training/training/core/dataset.py (hunk 3 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py:260]: nope it didn't

**Comment by codingwithsurya** [comment]
> > looks like one of the lints are failing. weird. @karkir0003 @DSGT-DLP/project-lead
> > the error message is this:
> > Note: This error originates from the build backend, and is likely not a problem with poetry but with simpleaudio (1.0.4) not supporting PEP 517 builds. You can verify this by running 'pip wheel --no-cache-dir --use-pep517 "simpleaudio (==1.0.4)"'.
> > Error: Process completed with exit code 1.
> > We just added torchaudio and soundata libraries in this pr to poetry so bc of that.
> 
> whats simpleaudio? is this a lib thats a dependency used by soundata?

still tryna figure this one out. i went through the logs and couldnt find any hints

**Comment by codingwithsurya** [comment]
> did we install the latest stable version of torchaudio and soundata?

yup i double checked. i just ran `dlp-cli backend add ____`

**Comment by codingwithsurya** [comment]
> > looks like one of the lints are failing. weird. @karkir0003 @DSGT-DLP/project-lead
> > the error message is this: Note: This error originates from the build backend, and is likely not a problem with poetry but with simpleaudio (1.0.4) not supporting PEP 517 builds. You can verify this by running 'pip wheel --no-cache-dir --use-pep517 "simpleaudio (==1.0.4)"'. Error: Process completed with exit code 1.
> > We just added torchaudio and soundata libraries in this pr to poetry so bc of that.
> 
> I'd try starting the debugging with the following:
> 
> 1. Try running the command that's shown in the log. See if any further logs come up
> 2. Maybe we might need to find a compatible version of `soundata` or potentially add `simpleaudio` as a dependency just like how you used DLP CLI to install `torchaudio`
> 3. If 1 and 2 don't work, try asking in the poetry github repo by filing a github issue. There's also a discord server for Poetry that can help clarify

alright bet sounds good

File: training/training/core/dataset.py (hunk 1 above)
Comment (karkir0003) [reply training/training/core/dataset.py:249]: ok. might want to clarify that

File: training/training/core/dataset.py (hunk 2 above)
Comment (karkir0003) [reply training/training/core/dataset.py:240]: got it. maybe we should have a utility preprocess function that stores these steps so that anyone who wants to learn more can better follow the logic

Comment (karkir0003) [reply training/training/core/dataset.py:240]: by this I mean a helper preprocess function for this specific dataset class you created

File: training/training/core/dataset.py (hunk 3 above)
Comment (karkir0003) [reply training/training/core/dataset.py:260]: sg. resolved

File: training/training/core/dataset.py (hunk 1 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py:249]: will do

File: training/training/core/dataset.py (hunk 2 above)
Comment (karkir0003) [reply training/training/core/dataset.py:240]: update from convo: Maybe for now, we don't need a custom preprocess function since we are onboarding one custom dataset with its own logic. If we do introduce new custom datasets beyond UrbanSound, maybe having a helper preprocess function for each custom dataset we onboard might help with readability.

**Comment by karkir0003** [comment]
@codingwithsurya , looks like someone from poetry responded to this thread I created and shared with you in discord: https://github.com/orgs/python-poetry/discussions/9418

**Comment by codingwithsurya** [comment]
> https://github.com/orgs/python-poetry/discussions/9418

ok. we can add it as a dependency then.

**Comment by karkir0003** [comment]
> > https://github.com/orgs/python-poetry/discussions/9418
> 
> ok. we can add it as a dependency then.

let's give this a try and see if that works @codingwithsurya
